core-rust/state-manager/src/mempool/simple_mempool.rs

```diff
@@ -196,52 +187,57 @@ impl SimpleMempool {
+    /// ASSUMPTION: Mempool does not already contain the transaction (panics otherwise).
+    /// Tries to add a new transaction into the mempool.
     /// Will return either a [`Vec`] of [`MempoolData`] that was evicted in order to fit the new transaction or an error
     /// if the mempool is full and the new transaction proposal priority is not better than what already exists.
     pub fn add_transaction(
         &mut self,
         transaction: Arc<MempoolTransaction>,
         source: MempoolAddSource,
+        added_at: Instant,
     ) -> Result<Vec<Arc<MempoolData>>, MempoolAddError> {
         let payload_hash = transaction.notarized_transaction_hash();
         let intent_hash = transaction.intent_hash();
-        let transaction_size = transaction.raw.0.len();
+        let transaction_size = transaction.raw.0.len() as u64;
 
-        let target_remaining_total_transactions_size = transaction_size;
-        let target_remaining_transaction_count = 1;
-        let transaction_data = Arc::new(MempoolData::create(transaction, Instant::now(), source));
+        let transaction_data = Arc::new(MempoolData::create(transaction, added_at, source));
         let new_order_data = MempoolDataProposalPriorityOrdering(transaction_data.clone());
 
-        let mut remaining_total_transactions_size = self.remaining_total_transactions_size;
-        let mut remaining_transaction_count = self.remaining_transaction_count;
+        let mut total_transaction_size_free_space = self.remaining_total_transactions_size;
+        let mut total_transaction_count_free_space = self.remaining_transaction_count;
         let mut to_be_removed = Vec::new();
         let mut priority_iter = self.proposal_priority_index.iter();
+        // Collect the lowest priority transactions that are required to be evicted in order to add the new one.
         // Note: worst-case scenario is the biggest transaction that will be rejected against a mempool full of smallest
         // possible transactions. This can be mitigated with a dynamic segment tree which can do fast range sum queries,
         // in order to check rejection before actually getting the evicted transactions. With a minimum transaction of
         // 1024 bytes and current max transaction size of 1MB this should not be a problem yet.
-        while remaining_total_transactions_size < target_remaining_total_transactions_size
-            || remaining_transaction_count < target_remaining_transaction_count
+        while total_transaction_size_free_space < transaction_size
+            || total_transaction_count_free_space < 1
         {
-            let order_data = priority_iter.next_back();
-            match order_data {
+            let lowest_priority_transaction = priority_iter.next();
+            match lowest_priority_transaction {
                 None => {
+                    // Even with an empty mempool we are not able to fulfill the request.
```

Comment:
Hmm I feel we should output something like this in this case, rather than panicking? Some of our tests configure a zero sized mempool, so would be a bit weird to panic.
```
    return Err(MempoolAddError::PriorityThresholdNotMet {
        min_tip_percentage_required: None,
        tip_percentage: transaction_data.transaction.tip_percentage(),
    });
```